ontotext-yasgui-web-component/src/services/translation.service.ts

```diff
@@ -0,0 +1,93 @@
+import en from '../i18n/locale-en.json'
+import fr from '../i18n/locale-fr.json'
+import {DEFAULT_LANG} from '../configurations/constants';
+
+class TranslationServiceDefinition {
+
+  private currentLang = DEFAULT_LANG;
+
+  private bundle = {en, fr}
+
+  setLanguage(lang: string) {
+    this.currentLang = lang;
+  }
+
+  /**
+   * Merges external i18e configuration with bundled one.
+   * * @param translations - external i18e configuration.
+   * Key -> value translations as JSON. If the language is supported, then not needed to pass all label values.
+   * If pass a new language then all label's values have to be present, otherwise they will be translated to the default English language.
+   * Example:
+   * {
+   *   en: {
+   *     "tooltip.switch.orientation.horizontal": "Switch to horizontal view",
+   *     "tooltip.switch.orientation.vertical": "Switch to vertical view",
+   *     "btn.mode-yasqe": "Editor only",
+   *     "btn.mode-yasgui": "Editor and results",
+   *     "btn.mode-yasr": "Results only",
+   *   }
+   *   fr: {
+   *     "tooltip.switch.orientation.vertical": "Basculer vers verticale voir",
+   *     "btn.mode-yasqe": "Éditeur seulement",
+   *   }
+   * }
+   */
+  addTranslations(translations: { locale: { key: string, value: string } }) {
+    Object.keys(translations).forEach((language) => {
+      // If is new language add it all
+      if (!this.bundle[language]) {
+        this.bundle[language] = translations[language];
+      } else {
+        // merge passed key -> values
+        Object.assign(this.bundle[language], translations[language]);
+      }
+    });
+  }
+
+  translate(key: string, parameters?: TranslationParameter[]): string {
+    let translation = this.getTranslation(key);
+    translation = this.applyParameters(translation, parameters);
+    return translation ? translation : key;
+  }
+
+  private getTranslation(key: string): string {
+    let selectedLang = this.currentLang;
+    if (!this.bundle || !this.bundle[this.currentLang]) {
+      console.warn('Missing locale file for [' + this.currentLang + ']');
+      selectedLang = DEFAULT_LANG;
+    }
+
+    let translation = this.bundle[selectedLang][key];
+    if (!translation) {
+      // Fallback to English
+      translation = this.bundle[DEFAULT_LANG][key];
```

Comment:
It's not guaranteed that the key will be present in the default language bundle, so we should return the key at the end.

Reply:
Yes, you are right, but this check is done at the end of  "translate" function.

Reply:
hmm I now saw that you actually return the key in the `translate` method if no translation is returned here. I think it'd be cleaner if you do it here. Also the parameters can be applied here too. This would simplify the logic and instant checking if translation is present